compiler/src/dotty/tools/dotc/core/Definitions.scala

```diff
@@ -1109,23 +1109,48 @@ class Definitions {
     sym.owner.linkedClass.typeRef
 
   object FunctionOf {
+    /** Matches a `FunctionN[...]`/`ContextFunctionN[...]` or refined `PolyFunction`/`FunctionN[...]`/`ContextFunctionN[...]`.
+     *  Extracts the method type type and apply info.
+     */
+    def unapply(ft: Type)(using Context): Option[MethodOrPoly] = {
+      ft match
+        case RefinedFunctionOf(mt) => Some(mt)
+        case FunctionNOf(argTypes, resultType, isContextual) =>
+          val methodType = if isContextual then ContextualMethodType else MethodType
+          Some(methodType(argTypes, resultType))
```

Comment:
Followup in #18443

Reply:
I agree it's good to make the conversion explicit in the name. The performance aspects are probably OK since the alternative is usually to take the list of AppliedType arguments and split it into inits and last, which is also expensive.